CDI 2.0 for WildFly 12

manovotn opened the pull request:
Adapt Wildfly 12 for Weld 3.
Implement new method in WeldSecurityServices. Enable Elytron, use PrivilegedActions.
As per WELD-1914, we need other modules to be able to use Weld API module. Hence we export it from the core module.

manovotn committed:
WFLY-6465 CDI 2.0 Support
Adapt Wildfly 12 for Weld 3.
Implement new method in WeldSecurityServices. Enable Elytron, use PrivilegedActions.
As per WELD-1914, we need other modules to be able to use Weld API module. Hence we export it from the core module.

manovotn commented:
We ran TCK tests of Weld 3 against CDI 1.2 TCK, the results can be seen [here](https://weld-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/Weld%203%20CI/job/Weld-3.x-WFLY-TCK-1.2/). There should only be one failure which is expected (and will appear in official TCK exclude list soon, based on [CDITCK-497](https://issues.jboss.org/browse/CDITCK-497) and respectively [CDI-545](https://issues.jboss.org/browse/CDI-545)).

jmesnil commented on feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/weld/main/module.xml:
```diff
@@ -37,6 +37,7 @@
         <module name="org.jboss.weld.core" />
         <module name="org.jboss.weld.spi" />
         <module name="org.jboss.as.weld.spi" />
+        <module name="org.picketbox" />
```
why is a dependency to PicketBox required?

manovotn replied:
I think this is because in Weld 3 you have async events and there we need to propagate security context. Hence `WeldSecurityServices` need to operate on classes from picketbox (or elytron, whichever is present) such as [`SecurityContextAssociation`](https://github.com/wildfly/wildfly/pull/10642/files#diff-a03cbe645209f431d4aab4d8f8ecc510R41).

ctomc replied:
i would say elytron is the main thing here, @darranl wdyt?

manovotn replied:
Not having this dependency will blow up with CNFE (on `org.jboss.security.SecurityContextAssociation`) when using async events. The reason is as I described above - we need to propagate security context and for that we need this dependency as the case can be that elytron isn't actually used (in which case there is going to be picketbox, right?).

That's at least from what I understood about elytron/picketbox, feel free to correct me.

manovotn commented:
@jmesnil the `Linux-elytron` CI build failed due to build timeout but there are no visible failures. I can see that the `Linux` build took even longer and did not timeout. Could you please look the timeout settings and re-run that CI?

jmesnil commented:
retest this please

jmesnil commented:
@manovotn The elytron run timed out again. Have you tried to run it locally (using `-Delytron`) to confirm it passes for you?

manovotn commented:
@jmesnil not the complete test suite (e.g. without -DallTests) but otherwise I did. Is there a doc on what switches/settings do I need to run complete suit? I can try that too if needed.
Although checking on the TC log, I cannot see any test failure there. Dunno what takes so long there.

manovotn commented:
@jmesnil I ran the TS locally using `mvn clean verify -DallTests -DfailIfNoTests=false -Dorg.jboss.model.test.cache.strict=true -Djboss.test.transformers.eap -Delytron` (copied over from how the CI job runs).

The build and all tests passed.

jmesnil commented:
retest this please

manovotn commented:
The suite now ran in time (91 min) but one test (marked as Flaky by TeamCity) failed with SocketException. The very same test succeeded in previous runs - unstable test/networking problems during execution perhaps?

jmesnil commented:
retest this please

stuartwdouglas commented on weld/subsystem/src/main/java/org/jboss/as/weld/util/PrivilegedActions.java:
```diff
@@ -0,0 +1,53 @@
+ * @author Jozef Hartinger
+ *
+ */
+public class PrivilegedActions {
```
Do we really need this? It seems to be a class that exists only to remove a cast, and now that we are starting to really try and reduce our runtime memory footprint this seems like it may be unnecessary?

mkouba replied:
It should remove the ambiguity between `PrivilegedAction` and `PrivilegedExceptionAction` which occurs when a lambda is passed to `AccessController`.

stuartwdouglas replied:
Yes, but you can do that explicitly via a cast which will not add anything to class loading time or runtime memory usage:

 (AccessController.doPrivileged((PrivilegedAction<Integer>) () -> 1);)

mkouba replied:
Ok, let's do it this way. On the other hand, such a small class does not affect the runtime memory usage significantly. I mean there are most probably better candidates in WildFly codebase and dependencies.

stuartwdouglas commented on feature-pack/src/main/resources/modules/system/layers/base/org/jboss/weld/core/main/module.xml:
```diff
@@ -30,6 +30,9 @@
 
     <resources>
         <artifact name="${org.jboss.weld:weld-core-impl}"/>
+        <artifact name="${org.jboss.weld.module:weld-ejb}"/>
```
Are there plans to move these into their own module so they can be excluded if not required (e.g. by swarm)?

mkouba replied:
Good point. We'll create a separate ticket to address this issue.

stuartwdouglas commented on feature-pack/src/main/resources/modules/system/layers/base/org/jboss/weld/core/main/module.xml:
```diff
@@ -45,8 +48,8 @@
         <module name="org.jboss.classfilewriter"/>
-        <module name="org.jboss.weld.api" />
-        <module name="org.jboss.weld.spi" />
+        <module name="org.jboss.weld.api" export="true"/>
```
Why have these been changed to be exported?

mkouba replied:
The reason was this issue: https://issues.jboss.org/browse/WELD-1914. All proxies/subclasses now implement `WeldConstruct` from API. But even so, I think it makes sense that any `weld-core-impl` consumer also depends on weld API and SPI.

stuartwdouglas reviewed (changes requested):
Needs a rebase

manovotn commented:
@stuartwdouglas sorry for the delay, I am currently on PTO. Rebased and replaced the util class with lambdas.

@mkouba thanks for answering the queries :)
